// ravendb/dist/commonjs/Documents/Indexes/StronglyTyped.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexingMapReduceFormatter = exports.IndexingReduceResults = exports.IndexingGroupResults = exports.StubMapUtils = void 0;
class StubMapUtils {
    load;
    cmpxchg;
    getMetadata;
    id;
    createSpatialField;
    createField;
    attachmentsFor;
    loadAttachment;
    loadAttachments;
    noTracking;
}
exports.StubMapUtils = StubMapUtils;
class IndexingGroupResults {
    groupBy(selector) {
        return new IndexingReduceResults(selector);
    }
}
exports.IndexingGroupResults = IndexingGroupResults;
class IndexingReduceResults {
    _group;
    constructor(selector) {
        this._group = selector;
    }
    aggregate(reducer) {
        return new IndexingMapReduceFormatter(this._group, reducer);
    }
}
exports.IndexingReduceResults = IndexingReduceResults;
class IndexingMapReduceFormatter {
    _groupBy;
    _aggregate;
    constructor(groupBy, aggregate) {
        this._groupBy = groupBy;
        this._aggregate = aggregate;
    }
    format() {
        return `groupBy(${this._groupBy}).aggregate(${this._aggregate})`;
    }
}
exports.IndexingMapReduceFormatter = IndexingMapReduceFormatter;
