// ravendb/dist/esm/Documents/Operations/DataArchival/ConfigureDataArchivalOperation.js
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class ConfigureDataArchivalOperation {
    _configuration;
    constructor(configuration) {
        this._configuration = configuration;
    }
    getCommand(conventions) {
        return new ConfigureDataArchivalCommand(conventions, this._configuration);
    }
    get resultType() {
        return "CommandResult";
    }
}
class ConfigureDataArchivalCommand extends RavenCommand {
    _conventions;
    _configuration;
    constructor(conventions, configuration) {
        super();
        this._conventions = conventions;
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/data-archival/config";
        const headers = this._headers()
            .typeAppJson()
            .build();
        const body = this._serializer.serialize(this._configuration);
        return {
            method: "POST",
            uri,
            headers,
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
