// ravendb/dist/esm/Documents/Operations/Backups/UpdatePeriodicBackupOperation.js
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class UpdatePeriodicBackupOperation {
    _configuration;
    constructor(configuration) {
        this._configuration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdatePeriodicBackupCommand(this._configuration);
    }
}
class UpdatePeriodicBackupCommand extends RavenCommand {
    _configuration;
    constructor(configuration) {
        super();
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/periodic-backup";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
