// ravendb/dist/esm/Documents/Commands/GetTcpInfoForRemoteTaskCommand.js
import { RavenCommand } from "../../Http/RavenCommand.js";
import { throwError } from "../../Exceptions/index.js";
export class GetTcpInfoForRemoteTaskCommand extends RavenCommand {
    _remoteDatabase;
    _remoteTask;
    _tag;
    _verifyDatabase;
    _requestedNode;
    constructor(tag, remoteDatabase, remoteTask, verifyDatabase = false) {
        super();
        if (!remoteDatabase) {
            throwError("InvalidArgumentException", "RemoteDatabase cannot be null");
        }
        this._remoteDatabase = remoteDatabase;
        if (!remoteTask) {
            throwError("InvalidArgumentException", "RemoteTask cannot be null");
        }
        this._remoteTask = remoteTask;
        this._tag = tag;
        this._verifyDatabase = verifyDatabase;
        this.timeout = 15_000;
    }
    createRequest(node) {
        let uri = node.url + "/info/remote-task/tcp?" +
            "database=" + this._urlEncode(this._remoteDatabase) +
            "&remote-task=" + this._urlEncode(this._remoteTask) +
            "&tag=" + this._urlEncode(this._tag);
        if (this._verifyDatabase) {
            uri += "&verify-database=true";
        }
        this._requestedNode = node;
        return {
            method: "GET",
            uri
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRequestedNode() {
        return this._requestedNode;
    }
    get isReadRequest() {
        return false;
    }
}
