// ravendb/dist/esm/ServerWide/Operations/OngoingTasks/PutServerWideExternalReplicationOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class PutServerWideExternalReplicationOperation {
    _configuration;
    constructor(configuration) {
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new PutServerWideExternalReplicationCommand(this._configuration, conventions);
    }
}
class PutServerWideExternalReplicationCommand extends RavenCommand {
    _configuration;
    constructor(configuration, conventions) {
        super();
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = conventions.objectMapper.toObjectLiteral(configuration);
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
    createRequest(node) {
        const uri = node.url + "/admin/configuration/server-wide/external-replication";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "PUT",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        return this._parseResponseDefaultAsync(bodyStream);
    }
}
