// ravendb/dist/esm/ServerWide/Operations/OngoingTasks/SetDatabasesLockOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
import { TypeUtil } from "../../../Utility/TypeUtil.js";
export class SetDatabasesLockOperation {
    _parameters;
    constructor(databaseNameOrParameters, mode) {
        if (!databaseNameOrParameters) {
            throwError("InvalidArgumentException", "Database or Parameters cannot be null");
        }
        if (TypeUtil.isString(databaseNameOrParameters)) {
            this._parameters = {
                databaseNames: [databaseNameOrParameters],
                mode
            };
        }
        else {
            if (!databaseNameOrParameters.databaseNames || databaseNameOrParameters.databaseNames.length === 0) {
                throwError("InvalidArgumentException", "DatabaseNames cannot be null or empty");
            }
            this._parameters = databaseNameOrParameters;
        }
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new SetDatabasesLockCommand(conventions, this._parameters);
    }
}
class SetDatabasesLockCommand extends RavenCommand {
    _parameters;
    constructor(conventions, parameters) {
        super();
        if (!conventions) {
            throwError("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!parameters) {
            throwError("InvalidArgumentException", "Parameters cannot be null");
        }
        this._parameters = conventions.objectMapper.toObjectLiteral(parameters);
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/set-lock";
        const body = this._serializer.serialize(this._parameters);
        const headers = this._headers()
            .typeAppJson().build();
        return {
            uri,
            method: "POST",
            body,
            headers
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
