// ravendb/dist/esm/Documents/Operations/Etl/ConnectionString.js
export class ConnectionString {
    name;
}
export class RavenConnectionString extends ConnectionString {
    database;
    topologyDiscoveryUrls;
    type = "Raven";
}
export class SqlConnectionString extends ConnectionString {
    connectionString;
    factoryName;
    type = "Sql";
}
export class OlapConnectionString extends ConnectionString {
    localSettings;
    s3Settings;
    azureSettings;
    glacierSettings;
    googleCloudSettings;
    ftpSettings;
    type = "Olap";
}
export class ElasticSearchConnectionString extends ConnectionString {
    nodes;
    authentication;
    /**
     * @deprecated Elasticsearch compatibility isn't required anymore to connect with Elasticsearch server v8.x.
     */
    enableCompatibilityMode;
    type = "ElasticSearch";
}
export class Authentication {
    apiKey;
    basic;
    certificate;
}
export class QueueConnectionString extends ConnectionString {
    brokerType;
    kafkaConnectionSettings;
    rabbitMqConnectionSettings;
    type = "Queue";
}
