// ravendb/dist/esm/Documents/Operations/Replication/PutPullReplicationAsHubOperation.js
import { StringUtil } from "../../../Utility/StringUtil.js";
import { throwError } from "../../../Exceptions/index.js";
import { TypeUtil } from "../../../Utility/TypeUtil.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class PutPullReplicationAsHubOperation {
    _pullReplicationDefinition;
    constructor(nameOrDefinition) {
        if (TypeUtil.isString(nameOrDefinition)) {
            const name = nameOrDefinition;
            if (StringUtil.isNullOrEmpty(name)) {
                throwError("InvalidArgumentException", "Name cannot be null or empty");
            }
            this._pullReplicationDefinition = {
                name
            };
        }
        else {
            const pullReplicationDefinition = nameOrDefinition;
            if (StringUtil.isNullOrEmpty(pullReplicationDefinition.name)) {
                throwError("InvalidArgumentException", "Name cannot be null or empty");
            }
            this._pullReplicationDefinition = pullReplicationDefinition;
        }
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdatePullReplicationDefinitionCommand(this._pullReplicationDefinition);
    }
}
class UpdatePullReplicationDefinitionCommand extends RavenCommand {
    _pullReplicationDefinition;
    constructor(pullReplicationDefinition) {
        super();
        this._pullReplicationDefinition = pullReplicationDefinition;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/tasks/pull-replication/hub";
        const body = this._serializer.serialize(this._pullReplicationDefinition);
        return {
            method: "PUT",
            uri,
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
