// ravendb/dist/commonjs/ServerWide/Operations/AddDatabaseNodeOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddDatabaseNodeOperation = void 0;
const index_js_1 = require("../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../Utility/RaftIdGenerator.js");
const TypeUtil_js_1 = require("../../Utility/TypeUtil.js");
const ClientShardHelper_js_1 = require("../../Utility/ClientShardHelper.js");
class AddDatabaseNodeOperation {
    _databaseName;
    _node;
    constructor(databaseName, nodeOrShardNumber, node) {
        if (!TypeUtil_js_1.TypeUtil.isNullOrUndefined(nodeOrShardNumber) && TypeUtil_js_1.TypeUtil.isNumber(nodeOrShardNumber)) {
            this._databaseName = ClientShardHelper_js_1.ClientShardHelper.toShardName(databaseName, nodeOrShardNumber);
            this._node = node;
        }
        else {
            this._databaseName = databaseName;
            this._node = nodeOrShardNumber;
        }
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new AddDatabaseNodeCommand(this._databaseName, this._node);
    }
}
exports.AddDatabaseNodeOperation = AddDatabaseNodeOperation;
class AddDatabaseNodeCommand extends RavenCommand_js_1.RavenCommand {
    _databaseName;
    _node;
    constructor(databaseName, node) {
        super();
        if (!databaseName) {
            (0, index_js_1.throwError)("InvalidArgumentException", "DatabaseName cannot be null");
        }
        this._databaseName = databaseName;
        this._node = node;
    }
    createRequest(node) {
        let uri = node.url + "/admin/databases/node?name=" + this._databaseName;
        if (this._node) {
            uri += "&node=" + this._node;
        }
        return {
            uri,
            method: "PUT"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
