// ravendb/dist/commonjs/ServerWide/Operations/PromoteDatabaseNodeOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromoteDatabaseNodeOperation = void 0;
const index_js_1 = require("../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../Utility/RaftIdGenerator.js");
const ClientShardHelper_js_1 = require("../../Utility/ClientShardHelper.js");
const TypeUtil_js_1 = require("../../Utility/TypeUtil.js");
class PromoteDatabaseNodeOperation {
    _databaseName;
    _node;
    constructor(databaseName, node, shardNumber) {
        this._node = node;
        this._databaseName = TypeUtil_js_1.TypeUtil.isNullOrUndefined(shardNumber) ? databaseName : ClientShardHelper_js_1.ClientShardHelper.toShardName(databaseName, shardNumber);
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new PromoteDatabaseNodeCommand(this._databaseName, this._node);
    }
}
exports.PromoteDatabaseNodeOperation = PromoteDatabaseNodeOperation;
class PromoteDatabaseNodeCommand extends RavenCommand_js_1.RavenCommand {
    _databaseName;
    _node;
    constructor(databaseName, node) {
        super();
        if (!databaseName) {
            (0, index_js_1.throwError)("InvalidArgumentException", "DatabaseName cannot be null");
        }
        if (!node) {
            (0, index_js_1.throwError)("InvalidArgumentException", "Node cannot be null");
        }
        this._databaseName = databaseName;
        this._node = node;
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/promote?name=" + this._databaseName + "&node=" + this._node;
        return {
            uri,
            method: "POST"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
