// ravendb/dist/commonjs/ServerWide/Operations/SetDatabaseDynamicDistributionOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetDatabaseDynamicDistributionOperation = void 0;
const StringUtil_js_1 = require("../../Utility/StringUtil.js");
const index_js_1 = require("../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../Utility/RaftIdGenerator.js");
class SetDatabaseDynamicDistributionOperation {
    _allowDynamicDistribution;
    _databaseName;
    constructor(databaseName, allowDynamicDistribution) {
        if (StringUtil_js_1.StringUtil.isNullOrEmpty(databaseName)) {
            (0, index_js_1.throwError)("InvalidArgumentException", "DatabaseName should not be null or empty");
        }
        this._allowDynamicDistribution = allowDynamicDistribution;
        this._databaseName = databaseName;
    }
    getCommand(conventions) {
        return new SetDatabaseDynamicDistributionCommand(this._databaseName, this._allowDynamicDistribution);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.SetDatabaseDynamicDistributionOperation = SetDatabaseDynamicDistributionOperation;
class SetDatabaseDynamicDistributionCommand extends RavenCommand_js_1.RavenCommand {
    _databaseName;
    _allowDynamicDistribution;
    constructor(databaseName, allowDynamicDistribution) {
        super();
        this._databaseName = databaseName;
        this._allowDynamicDistribution = allowDynamicDistribution;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/dynamic-node-distribution?name=" + this._databaseName + "&enabled=" + this._allowDynamicDistribution;
        return {
            uri,
            method: "POST"
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
