// ravendb/dist/commonjs/ServerWide/Operations/UpdateUnusedDatabasesOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateUnusedDatabasesOperation = void 0;
const StringUtil_js_1 = require("../../Utility/StringUtil.js");
const index_js_1 = require("../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../Utility/RaftIdGenerator.js");
class UpdateUnusedDatabasesOperation {
    _database;
    _parameters;
    constructor(database, unusedDatabaseIds, validate = false) {
        if (StringUtil_js_1.StringUtil.isNullOrEmpty(database)) {
            (0, index_js_1.throwError)("InvalidArgumentException", "Database cannot be null");
        }
        this._database = database;
        this._parameters = {
            databaseIds: unusedDatabaseIds,
            validate
        };
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdateUnusedDatabasesCommand(this._database, this._parameters);
    }
}
exports.UpdateUnusedDatabasesOperation = UpdateUnusedDatabasesOperation;
class UpdateUnusedDatabasesCommand extends RavenCommand_js_1.RavenCommand {
    _database;
    _parameters;
    constructor(database, parameters) {
        super();
        this._database = database;
        this._parameters = parameters;
    }
    createRequest(node) {
        let uri = node.url + "/admin/databases/unused-ids?name=" + this._database;
        if (this._parameters.validate) {
            uri += "&validate=true";
        }
        const body = this._serializer.serialize(this._parameters);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
