// ravendb/dist/commonjs/Documents/Queries/Suggestions/SuggestionQueryBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuggestionQueryBase = void 0;
const Stopwatch_js_1 = require("../../../Utility/Stopwatch.js");
const QueryCommand_js_1 = require("../../Commands/QueryCommand.js");
const LazySuggestionQueryOperation_js_1 = require("../../Session/Operations/Lazy/LazySuggestionQueryOperation.js");
const QueryOperation_js_1 = require("../../Session/Operations/QueryOperation.js");
const ObjectUtil_js_1 = require("../../../Utility/ObjectUtil.js");
class SuggestionQueryBase {
    _session;
    _query;
    _duration;
    constructor(session) {
        this._session = session;
    }
    async execute() {
        const command = this._getCommand();
        this._duration = Stopwatch_js_1.Stopwatch.createStarted();
        this._session.incrementRequestCount();
        await this._session.requestExecutor.execute(command);
        return this._processResults(command.result);
    }
    _processResults(queryResult) {
        this._invokeAfterQueryExecuted(queryResult);
        const results = {};
        for (const result of queryResult.results) {
            const transformedResult = ObjectUtil_js_1.ObjectUtil.transformObjectKeys(result, {
                defaultTransform: ObjectUtil_js_1.ObjectUtil.camel
            });
            results[transformedResult.name] = transformedResult;
        }
        QueryOperation_js_1.QueryOperation.ensureIsAcceptable(queryResult, this._query.waitForNonStaleResults, this._duration, this._session);
        return results;
    }
    executeLazy() {
        this._query = this._getIndexQuery();
        return this._session.addLazyOperation(new LazySuggestionQueryOperation_js_1.LazySuggestionQueryOperation(this._session, this._query, result => this._invokeAfterQueryExecuted(result), (result) => this._processResults(result)));
    }
    _getCommand() {
        this._query = this._getIndexQuery();
        return new QueryCommand_js_1.QueryCommand(this._session, this._query, {
            indexEntriesOnly: false,
            metadataOnly: false
        });
    }
    toString() {
        return this._getIndexQuery(false).toString();
    }
}
exports.SuggestionQueryBase = SuggestionQueryBase;
