// ravendb/dist/esm/ServerWide/Operations/DocumentsCompression/UpdateDocumentsCompressionConfigurationOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class UpdateDocumentsCompressionConfigurationOperation {
    _documentsCompressionConfiguration;
    constructor(configuration) {
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._documentsCompressionConfiguration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdateDocumentCompressionConfigurationCommand(this._documentsCompressionConfiguration);
    }
}
class UpdateDocumentCompressionConfigurationCommand extends RavenCommand {
    _documentsCompressionConfiguration;
    constructor(configuration) {
        super();
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._documentsCompressionConfiguration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/documents-compression/config";
        const headers = this._headers()
            .typeAppJson().build();
        const body = this._serializer.serialize(this._documentsCompressionConfiguration);
        return {
            uri,
            method: "POST",
            headers,
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            return;
        }
        return await this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
