// ravendb/dist/commonjs/Documents/Operations/QueueSink/UpdateQueueSinkOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateQueueSinkOperation = void 0;
const RavenCommand_js_1 = require("../../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../../Utility/RaftIdGenerator.js");
class UpdateQueueSinkOperation {
    _taskId;
    _configuration;
    constructor(taskId, configuration) {
        this._taskId = taskId;
        this._configuration = configuration;
    }
    getCommand(conventions) {
        return new UpdateQueueSinkCommand(conventions, this._taskId, this._configuration);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.UpdateQueueSinkOperation = UpdateQueueSinkOperation;
class UpdateQueueSinkCommand extends RavenCommand_js_1.RavenCommand {
    _taskId;
    _configuration;
    _conventions;
    constructor(conventions, taskId, configuration) {
        super();
        this._taskId = taskId;
        this._configuration = configuration;
        this._conventions = conventions;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/queue-sink?id=" + this._taskId;
        const headers = this._headers().typeAppJson().build();
        const body = this._serializer.serialize(this._configuration);
        return {
            method: "PUT",
            uri,
            body,
            headers
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
