// ravendb/dist/commonjs/ServerWide/Sharding/AddDatabaseShardOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddDatabaseShardOperation = void 0;
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const TypeUtil_js_1 = require("../../Utility/TypeUtil.js");
const RaftIdGenerator_js_1 = require("../../Utility/RaftIdGenerator.js");
class AddDatabaseShardOperation {
    _databaseName;
    _shardNumber;
    _nodes;
    _replicationFactor;
    _dynamicNodeDistribution;
    constructor(parameters) {
        this._databaseName = parameters.databaseName;
        this._shardNumber = parameters.shardNumber;
        this._nodes = parameters.nodes;
        this._replicationFactor = parameters.replicationFactor;
        this._dynamicNodeDistribution = parameters.dynamicNodeDistribution;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new AddDatabaseShardCommand(this._databaseName, this._shardNumber, this._nodes, this._replicationFactor, this._dynamicNodeDistribution);
    }
}
exports.AddDatabaseShardOperation = AddDatabaseShardOperation;
class AddDatabaseShardCommand extends RavenCommand_js_1.RavenCommand {
    _databaseName;
    _shardNumber;
    _nodes;
    _replicationFactor;
    _dynamicNodeDistribution;
    constructor(databaseName, shardNumber, nodes, replicationFactor, dynamicNodeDistribution) {
        super();
        this._databaseName = databaseName;
        this._shardNumber = shardNumber;
        this._nodes = nodes;
        this._replicationFactor = replicationFactor;
        this._dynamicNodeDistribution = dynamicNodeDistribution;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        let uri = node.url + "/admin/databases/shard?name=" + this._urlEncode(this._databaseName);
        if (!TypeUtil_js_1.TypeUtil.isNullOrUndefined(this._shardNumber)) {
            uri += "&shardNumber=" + this._shardNumber;
        }
        if (!TypeUtil_js_1.TypeUtil.isNullOrUndefined(this._replicationFactor)) {
            uri += "&replicationFactor=" + this._replicationFactor;
        }
        if (!TypeUtil_js_1.TypeUtil.isNullOrUndefined(this._dynamicNodeDistribution)) {
            uri += "&dynamicNodeDistribution=" + this._dynamicNodeDistribution;
        }
        if (this._nodes && this._nodes.length > 0) {
            for (const nodeStr of this._nodes) {
                uri += "&node=" + this._urlEncode(nodeStr);
            }
        }
        return {
            uri,
            method: "PUT"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
