// ravendb/dist/commonjs/ServerWide/Commands/GetDatabaseTopologyCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetDatabaseTopologyCommand = void 0;
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const ServerNode_js_1 = require("../../Http/ServerNode.js");
const Topology_js_1 = require("../../Http/Topology.js");
const ObjectUtil_js_1 = require("../../Utility/ObjectUtil.js");
class GetDatabaseTopologyCommand extends RavenCommand_js_1.RavenCommand {
    _applicationIdentifier;
    _debugTag;
    constructor(debugTag, applicationIdentifier) {
        super();
        this._debugTag = debugTag;
        this._applicationIdentifier = applicationIdentifier;
        this.timeout = 15_000;
    }
    createRequest(node) {
        let uri = `${node.url}/topology?name=${node.database}`;
        if (node.url.toLowerCase().includes(".fiddler")) {
            // we want to keep the '.fiddler' stuff there so we'll keep tracking request
            // so we are going to ask the server to respect it
            uri += "&localUrl=" + encodeURIComponent(node.url);
        }
        if (this._debugTag) {
            uri += "&" + this._debugTag;
        }
        if (this._applicationIdentifier) {
            uri += "&applicationIdentifier=" + this._urlEncode(this._applicationIdentifier);
        }
        return { uri };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            return;
        }
        let body = null;
        const rawTpl = await this._pipeline()
            .collectBody(_ => body = _)
            .parseJsonSync()
            .objectKeysTransform(ObjectUtil_js_1.ObjectUtil.camel)
            .process(bodyStream);
        const nodes = rawTpl.nodes
            ? rawTpl.nodes.map(x => Object.assign(new ServerNode_js_1.ServerNode(), x))
            : null;
        this.result = new Topology_js_1.Topology(rawTpl.etag, nodes);
        return body;
    }
    get isReadRequest() {
        return true;
    }
}
exports.GetDatabaseTopologyCommand = GetDatabaseTopologyCommand;
