// ravendb/dist/commonjs/Documents/Operations/Replication/UnregisterReplicationHubAccessOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnregisterReplicationHubAccessOperation = void 0;
const StringUtil_js_1 = require("../../../Utility/StringUtil.js");
const index_js_1 = require("../../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../../Utility/RaftIdGenerator.js");
class UnregisterReplicationHubAccessOperation {
    _hubName;
    _thumbprint;
    constructor(hubName, thumbprint) {
        if (StringUtil_js_1.StringUtil.isNullOrEmpty(hubName)) {
            (0, index_js_1.throwError)("InvalidArgumentException", "HubName cannot be null or whitespace");
        }
        if (StringUtil_js_1.StringUtil.isNullOrEmpty(thumbprint)) {
            (0, index_js_1.throwError)("InvalidArgumentException", "Thumbprint cannot be null or whitespace.");
        }
        this._hubName = hubName;
        this._thumbprint = thumbprint;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UnregisterReplicationHubAccessCommand(this._hubName, this._thumbprint);
    }
}
exports.UnregisterReplicationHubAccessOperation = UnregisterReplicationHubAccessOperation;
class UnregisterReplicationHubAccessCommand extends RavenCommand_js_1.RavenCommand {
    _hubName;
    _thumbprint;
    constructor(hubName, thumbprint) {
        super();
        this._hubName = hubName;
        this._thumbprint = thumbprint;
        this._responseType = "Empty";
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database
            + "/admin/tasks/pull-replication/hub/access?name=" + this._urlEncode(this._hubName)
            + "&thumbprint=" + this._urlEncode(this._thumbprint);
        return {
            uri,
            method: "DELETE"
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
