// ravendb/dist/esm/Documents/Operations/TimeSeries/ConfigureTimeSeriesOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
export class ConfigureTimeSeriesOperation {
    _configuration;
    constructor(configuration) {
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new ConfigureTimeSeriesCommand(this._configuration);
    }
}
class ConfigureTimeSeriesCommand extends RavenCommand {
    _configuration;
    constructor(configuration) {
        super();
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/timeseries/config";
        const body = this._serializer.serialize(this._configuration.serialize());
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
