// ravendb/dist/esm/Documents/Operations/TimeSeries/GetTimeSeriesStatisticsOperation.js
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { DateUtil } from "../../../Utility/DateUtil.js";
export class GetTimeSeriesStatisticsOperation {
    _documentId;
    /**
     * Retrieve start, end and total number of entries for all time-series of a given document
     * @param documentId Document id
     */
    constructor(documentId) {
        this._documentId = documentId;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(store, conventions, httpCache) {
        return new GetTimeSeriesStatisticsCommand(conventions, this._documentId);
    }
}
class GetTimeSeriesStatisticsCommand extends RavenCommand {
    _conventions;
    _documentId;
    constructor(conventions, documentId) {
        super();
        this._conventions = conventions;
        this._documentId = documentId;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/timeseries/stats?docId=" + this._urlEncode(this._documentId);
        return {
            method: "GET",
            uri
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        let body = null;
        if (!bodyStream) {
            this.result = null;
            return null;
        }
        const results = await this._defaultPipeline(_ => body = _).process(bodyStream);
        const { timeSeries, ...restProps } = results;
        this.result = {
            ...restProps,
            timeSeries: timeSeries.map(t => {
                const { startDate, endDate } = t;
                return {
                    ...t,
                    startDate: DateUtil.utc.parse(startDate),
                    endDate: DateUtil.utc.parse(endDate)
                };
            })
        };
        return body;
    }
}
