// ravendb/dist/commonjs/Documents/Queries/FilterFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterFactory = void 0;
class FilterFactory {
    _documentQuery;
    constructor(documentQuery, filterLimit = Number.MAX_SAFE_INTEGER) {
        this._documentQuery = documentQuery;
        this._setFilterLimit(filterLimit);
    }
    equals(fieldNameOrWhereParams, value, exact = false) {
        this._documentQuery._whereEquals(fieldNameOrWhereParams, value, exact);
        return this;
    }
    notEquals(fieldNameOrWhereParams, value, exact = false) {
        this._documentQuery._whereNotEquals(fieldNameOrWhereParams, value, exact);
        return this;
    }
    greaterThan(fieldName, value) {
        this._documentQuery._whereGreaterThan(fieldName, value);
        return this;
    }
    greaterThanOrEqual(fieldName, value) {
        this._documentQuery._whereGreaterThanOrEqual(fieldName, value);
        return this;
    }
    lessThan(fieldName, value) {
        this._documentQuery._whereLessThan(fieldName, value);
        return this;
    }
    lessThanOrEqual(fieldName, value) {
        this._documentQuery._whereLessThanOrEqual(fieldName, value);
        return this;
    }
    andAlso() {
        this._documentQuery._andAlso();
        return this;
    }
    orElse() {
        this._documentQuery._orElse();
        return this;
    }
    not() {
        this._documentQuery._negateNext();
        return this;
    }
    openSubclause() {
        this._documentQuery._openSubclause();
        return this;
    }
    closeSubclause() {
        this._documentQuery._closeSubclause();
        return this;
    }
    _setFilterLimit(limit) {
        this._documentQuery._addFilterLimit(limit);
    }
}
exports.FilterFactory = FilterFactory;
