// ravendb/dist/commonjs/Documents/Operations/GetDetailedStatisticsOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DetailedDatabaseStatisticsCommand = exports.GetDetailedStatisticsOperation = void 0;
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
class GetDetailedStatisticsOperation {
    _debugTag;
    constructor(debugTag) {
        this._debugTag = debugTag;
    }
    getCommand(conventions) {
        return new DetailedDatabaseStatisticsCommand(this._debugTag);
    }
    get resultType() {
        return "CommandResult";
    }
}
exports.GetDetailedStatisticsOperation = GetDetailedStatisticsOperation;
class DetailedDatabaseStatisticsCommand extends RavenCommand_js_1.RavenCommand {
    _debugTag;
    constructor(debugTag) {
        super();
        this._debugTag = debugTag;
    }
    createRequest(node) {
        let uri = node.url + "/databases/" + node.database + "/stats/detailed";
        if (this._debugTag) {
            uri += "?" + this._debugTag;
        }
        return { uri };
    }
    async setResponseAsync(bodyStream, fromCache) {
        return await this._parseResponseDefaultAsync(bodyStream);
    }
    get isReadRequest() {
        return false;
    }
}
exports.DetailedDatabaseStatisticsCommand = DetailedDatabaseStatisticsCommand;
