// ravendb/dist/commonjs/Documents/Smuggler/DatabaseSmugglerOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseSmugglerOptions = void 0;
class DatabaseSmugglerOptions {
    static DEFAULT_OPERATE_ON_TYPES = [
        "Indexes", "Documents", "RevisionDocuments", "Conflicts", "DatabaseRecord", "ReplicationHubCertificates", "Identities",
        "CompareExchange", "Attachments", "CounterGroups", "Subscriptions", "TimeSeries"
    ];
    static DEFAULT_OPERATE_ON_DATABASE_RECORD_TYPES = [
        "Client", "ConflictSolverConfig", "Expiration", "ExternalReplications", "PeriodicBackups", "RavenConnectionStrings",
        "RavenEtls", "Revisions", "Settings", "SqlConnectionStrings", "Sorters", "SqlEtls",
        "HubPullReplications", "SinkPullReplications", "TimeSeries", "DocumentsCompression",
        "Analyzers", "LockMode", "OlapConnectionStrings", "OlapEtls", "ElasticSearchConnectionStrings",
        "ElasticSearchEtls", "PostgreSqlIntegration", "QueueConnectionStrings", "QueueEtl",
        "IndexesHistory", "Refresh", "QueueSinks", "DataArchival"
    ];
    static DEFAULT_MAX_STEPS_FOR_TRANSFORM_SCRIPT = 10 * 1_000;
    operateOnTypes;
    operateOnDatabaseRecordType;
    includeExpired;
    includeArtificial;
    includeArchived;
    removeAnalyzers;
    transformScript;
    maxStepsForTransformScript;
    skipRevisionCreation;
    encryptionKey;
    collections;
    /**
     * In case the database is corrupted (for example, Compression Dictionaries are lost), it is possible to export all the remaining data.
     */
    skipCorruptedData;
    constructor() {
        this.operateOnTypes = [...DatabaseSmugglerOptions.DEFAULT_OPERATE_ON_TYPES];
        this.operateOnDatabaseRecordType = [...DatabaseSmugglerOptions.DEFAULT_OPERATE_ON_DATABASE_RECORD_TYPES];
        this.maxStepsForTransformScript = DatabaseSmugglerOptions.DEFAULT_MAX_STEPS_FOR_TRANSFORM_SCRIPT;
        this.includeExpired = true;
        this.includeArchived = true;
        this.collections = [];
    }
}
exports.DatabaseSmugglerOptions = DatabaseSmugglerOptions;
