// ravendb/dist/commonjs/Documents/Operations/Analyzers/DeleteAnalyzerOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteAnalyzerOperation = void 0;
const index_js_1 = require("../../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../../Utility/RaftIdGenerator.js");
class DeleteAnalyzerOperation {
    _analyzerName;
    constructor(analyzerName) {
        if (!analyzerName) {
            (0, index_js_1.throwError)("InvalidArgumentException", "AnalyzerName cannot be null");
        }
        this._analyzerName = analyzerName;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteAnalyzerCommand(this._analyzerName);
    }
}
exports.DeleteAnalyzerOperation = DeleteAnalyzerOperation;
class DeleteAnalyzerCommand extends RavenCommand_js_1.RavenCommand {
    _analyzerName;
    constructor(analyzerName) {
        super();
        if (!analyzerName) {
            (0, index_js_1.throwError)("InvalidArgumentException", "AnalyzerName cannot be null");
        }
        this._analyzerName = analyzerName;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/analyzers?name=" + encodeURIComponent(this._analyzerName);
        return {
            uri,
            method: "DELETE"
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
