// ravendb/dist/esm/Documents/Operations/Sorters/DeleteSorterOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class DeleteSorterOperation {
    _sorterName;
    constructor(sorterName) {
        if (!sorterName) {
            throwError("InvalidArgumentException", "SorterName cannot be null");
        }
        this._sorterName = sorterName;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteSorterCommand(this._sorterName);
    }
}
class DeleteSorterCommand extends RavenCommand {
    _sorterName;
    constructor(sorterName) {
        super();
        if (!sorterName) {
            throwError("InvalidArgumentException", "IndexName cannot be null");
        }
        this._sorterName = sorterName;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/sorters?name=" + encodeURIComponent(this._sorterName);
        return {
            uri,
            method: "DELETE"
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
