// ravendb/dist/esm/Documents/Operations/MaintenanceOperationExecutor.js
import { OperationCompletionAwaiter } from "./OperationCompletionAwaiter.js";
import { ServerOperationExecutor } from "./ServerOperationExecutor.js";
import { throwError } from "../../Exceptions/index.js";
import { TypeUtil } from "../../Utility/TypeUtil.js";
export class MaintenanceOperationExecutor {
    _store;
    _nodeTag;
    _shardNumber;
    _databaseName;
    _requestExecutor;
    _serverOperationExecutor;
    constructor(store, databaseName, nodeTag, shardNumber) {
        this._store = store;
        this._databaseName = databaseName || store.database;
        this._nodeTag = nodeTag;
        this._shardNumber = shardNumber;
    }
    get requestExecutor() {
        if (this._requestExecutor) {
            return this._requestExecutor;
        }
        this._requestExecutor = this._databaseName ? this._store.getRequestExecutor(this._databaseName) : null;
        return this.requestExecutor;
    }
    get server() {
        if (!this._serverOperationExecutor) {
            this._serverOperationExecutor = new ServerOperationExecutor(this._store);
        }
        return this._serverOperationExecutor;
    }
    forDatabase(databaseName) {
        if (this._databaseName
            && this._databaseName.toLowerCase() === (databaseName || "").toLowerCase()) {
            return this;
        }
        return new MaintenanceOperationExecutor(this._store, databaseName, this._nodeTag, this._shardNumber);
    }
    async send(operation) {
        this._assertDatabaseNameSet();
        const command = operation.getCommand(this.requestExecutor.conventions);
        this.applyNodeTagAndShardNumberToCommandIfSet(command);
        await this.requestExecutor.execute(command);
        if (operation.resultType === "OperationId") {
            const idResult = command.result;
            return new OperationCompletionAwaiter(this.requestExecutor, this.requestExecutor.conventions, idResult.operationId, command.selectedNodeTag || idResult.operationNodeTag);
        }
        return command.result;
    }
    _assertDatabaseNameSet() {
        if (!this._databaseName) {
            throwError("InvalidOperationException", "Cannot use maintenance without a database defined, did you forget to call forDatabase?");
        }
    }
    applyNodeTagAndShardNumberToCommandIfSet(command) {
        if (this._nodeTag) {
            command.selectedNodeTag = this._nodeTag;
        }
        if (!TypeUtil.isNullOrUndefined(this._shardNumber)) {
            command.selectedShardNumber = this._shardNumber;
        }
    }
}
