// ravendb/dist/commonjs/Documents/Operations/Backups/DelayBackupOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelayBackupOperation = void 0;
const RavenCommand_js_1 = require("../../../Http/RavenCommand.js");
const TimeUtil_js_1 = require("../../../Utility/TimeUtil.js");
class DelayBackupOperation {
    _runningBackupTaskId;
    _duration;
    constructor(runningBackupTaskId, duration) {
        this._runningBackupTaskId = runningBackupTaskId;
        this._duration = duration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DelayBackupCommand(this._runningBackupTaskId, this._duration);
    }
}
exports.DelayBackupOperation = DelayBackupOperation;
class DelayBackupCommand extends RavenCommand_js_1.RavenCommand {
    _taskId;
    _duration;
    constructor(taskId, duration) {
        super();
        this._taskId = taskId;
        this._duration = duration;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/admin/backup-task/delay?taskId=" + this._taskId + "&duration=" + TimeUtil_js_1.TimeUtil.durationToTimeSpan(this._duration) + "&database=" + node.database;
        return {
            uri,
            method: "POST"
        };
    }
}
