// ravendb/dist/commonjs/Documents/Operations/Backups/UpdatePeriodicBackupOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdatePeriodicBackupOperation = void 0;
const RavenCommand_js_1 = require("../../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../../Utility/RaftIdGenerator.js");
class UpdatePeriodicBackupOperation {
    _configuration;
    constructor(configuration) {
        this._configuration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new UpdatePeriodicBackupCommand(this._configuration);
    }
}
exports.UpdatePeriodicBackupOperation = UpdatePeriodicBackupOperation;
class UpdatePeriodicBackupCommand extends RavenCommand_js_1.RavenCommand {
    _configuration;
    constructor(configuration) {
        super();
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/periodic-backup";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
