// ravendb/dist/esm/Documents/Operations/ConnectionStrings/PutConnectionStringOperation.js
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
import { throwError } from "../../../Exceptions/index.js";
export class PutConnectionStringOperation {
    _connectionString;
    constructor(connectionString) {
        this._connectionString = connectionString;
    }
    getCommand(conventions) {
        return new PutConnectionStringCommand(this._connectionString);
    }
    get resultType() {
        return "CommandResult";
    }
}
export class PutConnectionStringCommand extends RavenCommand {
    _connectionString;
    constructor(connectionString) {
        super();
        if (!connectionString) {
            throwError("InvalidArgumentException", "ConnectionString cannot be null");
        }
        this._connectionString = connectionString;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/admin/connection-strings";
        const headers = this._headers()
            .typeAppJson()
            .build();
        const body = this._serializer.serialize(this._connectionString);
        return {
            method: "PUT",
            uri,
            headers,
            body
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
        }
        return this._parseResponseDefaultAsync(bodyStream);
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
