// ravendb/dist/esm/Documents/Operations/Indexes/DisableIndexOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class DisableIndexOperation {
    _indexName;
    _clusterWide;
    constructor(indexName, clusterWide = false) {
        if (!indexName) {
            throwError("InvalidArgumentException", "IndexName cannot be null");
        }
        this._indexName = indexName;
        this._clusterWide = clusterWide;
    }
    getCommand(conventions) {
        return new DisableIndexCommand(this._indexName, this._clusterWide);
    }
    get resultType() {
        return "CommandResult";
    }
}
export class DisableIndexCommand extends RavenCommand {
    get isReadRequest() {
        return false;
    }
    _indexName;
    _clusterWide;
    constructor(indexName, clusterWide) {
        super();
        if (!indexName) {
            throwError("InvalidArgumentException", "IndexName cannot be null");
        }
        this._responseType = "Empty";
        this._indexName = indexName;
        this._clusterWide = clusterWide;
    }
    createRequest(node) {
        const uri = node.url
            + "/databases/" + node.database
            + "/admin/indexes/disable?name=" + encodeURIComponent(this._indexName)
            + "&clusterWide=" + this._clusterWide;
        return { method: "POST", uri };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
