// ravendb/dist/esm/Documents/Operations/Indexes/ResetIndexOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
export class ResetIndexOperation {
    _indexName;
    constructor(indexName) {
        if (!indexName) {
            throwError("InvalidArgumentException", "Index name cannot be null.");
        }
        this._indexName = indexName;
    }
    getCommand(conventions) {
        return new ResetIndexCommand(this._indexName);
    }
    get resultType() {
        return "CommandResult";
    }
}
export class ResetIndexCommand extends RavenCommand {
    _indexName;
    constructor(indexName) {
        super();
        if (!indexName) {
            throwError("InvalidArgumentException", "Index name cannot be null.");
        }
        this._indexName = indexName;
        this._responseType = "Empty";
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/indexes?name=" + encodeURIComponent(this._indexName);
        return { method: "RESET", uri };
    }
    get isReadRequest() {
        return false;
    }
}
