// ravendb/dist/esm/ServerWide/Operations/Certificates/GetCertificatesMetadataOperation.js
import { StringUtil } from "../../../Utility/StringUtil.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
export class GetCertificatesMetadataOperation {
    _name;
    constructor(name) {
        this._name = name;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new GetCertificatesMetadataCommand(conventions, this._name);
    }
}
class GetCertificatesMetadataCommand extends RavenCommand {
    _conventions;
    _name;
    constructor(conventions, name) {
        super();
        this._conventions = conventions;
        this._name = name;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        let uri = node.url + "/admin/certificates?metadataOnly=true";
        if (!StringUtil.isNullOrEmpty(this._name)) {
            uri += "&name=" + encodeURIComponent(this._name);
        }
        return {
            uri,
            method: "GET"
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            return;
        }
        let body = null;
        const results = await this._defaultPipeline(_ => body = _).process(bodyStream);
        this.result = this._conventions.objectMapper.fromObjectLiteral(results, {
            nestedTypes: {
                "results[].notAfter": "date",
                "results[].notBefore": "date",
            }
        }).results;
        return body;
    }
}
