// ravendb/dist/commonjs/Documents/Commands/ConditionalGetDocumentsCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConditionalGetDocumentsCommand = void 0;
const RavenCommand_js_1 = require("../../Http/RavenCommand.js");
const StatusCode_js_1 = require("../../Http/StatusCode.js");
const StreamUtil_js_1 = require("../../Utility/StreamUtil.js");
const ObjectUtil_js_1 = require("../../Utility/ObjectUtil.js");
const Constants_js_1 = require("../../Constants.js");
class ConditionalGetDocumentsCommand extends RavenCommand_js_1.RavenCommand {
    _changeVector;
    _id;
    _conventions;
    constructor(id, changeVector, conventions) {
        super();
        this._changeVector = changeVector;
        this._id = id;
        this._conventions = conventions;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/docs?id=" + this._urlEncode(this._id);
        return {
            uri,
            method: "GET",
            headers: {
                [Constants_js_1.HEADERS.IF_NONE_MATCH]: `"${this._changeVector}"`
            }
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this.result = null;
            return;
        }
        let body = null;
        this.result =
            await ConditionalGetDocumentsCommand.parseDocumentsResultResponseAsync(bodyStream, this._conventions, b => body = b);
        return body;
    }
    static async parseDocumentsResultResponseAsync(bodyStream, conventions, bodyCallback) {
        const body = await (0, StreamUtil_js_1.readToEnd)(bodyStream);
        bodyCallback?.(body);
        const parsedJson = JSON.parse(body);
        return ConditionalGetDocumentsCommand._mapToLocalObject(parsedJson, conventions);
    }
    static _mapToLocalObject(json, conventions) {
        return {
            results: json.Results.map(x => ObjectUtil_js_1.ObjectUtil.transformDocumentKeys(x, conventions)),
            changeVector: json.ChangeVector
        };
    }
    async processResponse(cache, response, bodyStream, url) {
        if (response.status === StatusCode_js_1.StatusCodes.NotModified) {
            return "Automatic";
        }
        const result = await super.processResponse(cache, response, bodyStream, url);
        this.result.changeVector = response.headers.get("ETag");
        return result;
    }
    /**
     * Here we explicitly do _NOT_ want to have caching
     * by the Request Executor, we want to manage it ourselves
     */
    get isReadRequest() {
        return false;
    }
}
exports.ConditionalGetDocumentsCommand = ConditionalGetDocumentsCommand;
