// ravendb/dist/esm/ServerWide/Operations/GetDatabaseNamesOperation.js
import { RavenCommand } from "../../Http/RavenCommand.js";
export class GetDatabaseNamesOperation {
    _start;
    _pageSize;
    constructor(start, pageSize) {
        this._start = start;
        this._pageSize = pageSize;
    }
    getCommand(conventions) {
        return new GetDatabaseNamesCommand(this._start, this._pageSize);
    }
    get resultType() {
        return "CommandResult";
    }
}
export class GetDatabaseNamesCommand extends RavenCommand {
    _start;
    _pageSize;
    constructor(start, pageSize) {
        super();
        this._start = start;
        this._pageSize = pageSize;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = `${node.url}/databases?start=${this._start}&pageSize=${this._pageSize}&namesOnly=true`;
        return { uri };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            this._throwInvalidResponse();
            return;
        }
        let body = null;
        const results = await this._defaultPipeline(_ => body = _)
            .process(bodyStream);
        const { databases } = results;
        if (!databases || !Array.isArray(databases)) {
            this._throwInvalidResponse();
        }
        this.result = databases;
        return body;
    }
}
