// ravendb/dist/esm/ServerWide/Operations/ReorderDatabaseMembersOperation.js
import { throwError } from "../../Exceptions/index.js";
import { RavenCommand } from "../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../Utility/RaftIdGenerator.js";
export class ReorderDatabaseMembersOperation {
    _database;
    _parameters;
    constructor(database, order, fixed = false) {
        if (!order || order.length === 0) {
            throwError("InvalidArgumentException", "Order list must contain values");
        }
        this._database = database;
        this._parameters = {
            membersOrder: order,
            fixed
        };
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new ReorderDatabaseMembersCommand(this._database, this._parameters);
    }
}
class ReorderDatabaseMembersCommand extends RavenCommand {
    _databaseName;
    _parameters;
    constructor(databaseName, parameters) {
        super();
        if (!databaseName) {
            throwError("InvalidArgumentException", "Database cannot be empty");
        }
        this._databaseName = databaseName;
        this._parameters = parameters;
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/reorder?name=" + this._databaseName;
        const body = this._serializer.serialize(this._parameters);
        return {
            uri,
            method: "POST",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
