// ravendb/dist/esm/ServerWide/Operations/SetDatabaseDynamicDistributionOperation.js
import { StringUtil } from "../../Utility/StringUtil.js";
import { throwError } from "../../Exceptions/index.js";
import { RavenCommand } from "../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../Utility/RaftIdGenerator.js";
export class SetDatabaseDynamicDistributionOperation {
    _allowDynamicDistribution;
    _databaseName;
    constructor(databaseName, allowDynamicDistribution) {
        if (StringUtil.isNullOrEmpty(databaseName)) {
            throwError("InvalidArgumentException", "DatabaseName should not be null or empty");
        }
        this._allowDynamicDistribution = allowDynamicDistribution;
        this._databaseName = databaseName;
    }
    getCommand(conventions) {
        return new SetDatabaseDynamicDistributionCommand(this._databaseName, this._allowDynamicDistribution);
    }
    get resultType() {
        return "CommandResult";
    }
}
class SetDatabaseDynamicDistributionCommand extends RavenCommand {
    _databaseName;
    _allowDynamicDistribution;
    constructor(databaseName, allowDynamicDistribution) {
        super();
        this._databaseName = databaseName;
        this._allowDynamicDistribution = allowDynamicDistribution;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/admin/databases/dynamic-node-distribution?name=" + this._databaseName + "&enabled=" + this._allowDynamicDistribution;
        return {
            uri,
            method: "POST"
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
