// ravendb/dist/commonjs/Documents/Operations/Etl/ConnectionString.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueConnectionString = exports.Authentication = exports.ElasticSearchConnectionString = exports.OlapConnectionString = exports.SqlConnectionString = exports.RavenConnectionString = exports.ConnectionString = void 0;
class ConnectionString {
    name;
}
exports.ConnectionString = ConnectionString;
class RavenConnectionString extends ConnectionString {
    database;
    topologyDiscoveryUrls;
    type = "Raven";
}
exports.RavenConnectionString = RavenConnectionString;
class SqlConnectionString extends ConnectionString {
    connectionString;
    factoryName;
    type = "Sql";
}
exports.SqlConnectionString = SqlConnectionString;
class OlapConnectionString extends ConnectionString {
    localSettings;
    s3Settings;
    azureSettings;
    glacierSettings;
    googleCloudSettings;
    ftpSettings;
    type = "Olap";
}
exports.OlapConnectionString = OlapConnectionString;
class ElasticSearchConnectionString extends ConnectionString {
    nodes;
    authentication;
    /**
     * @deprecated Elasticsearch compatibility isn't required anymore to connect with Elasticsearch server v8.x.
     */
    enableCompatibilityMode;
    type = "ElasticSearch";
}
exports.ElasticSearchConnectionString = ElasticSearchConnectionString;
class Authentication {
    apiKey;
    basic;
    certificate;
}
exports.Authentication = Authentication;
class QueueConnectionString extends ConnectionString {
    brokerType;
    kafkaConnectionSettings;
    rabbitMqConnectionSettings;
    type = "Queue";
}
exports.QueueConnectionString = QueueConnectionString;
