// ravendb/dist/esm/Documents/Commands/MultiGet/GetResponse.js
import { StatusCodes } from "../../../Http/StatusCode.js";
export class GetResponse {
    constructor() {
        this._headers = {};
    }
    _elapsed;
    _result;
    _headers;
    _statusCode;
    _forceRetry;
    static create(data) {
        return Object.assign(new GetResponse(), data);
    }
    get elapsed() {
        return this._elapsed;
    }
    set elapsed(elapsed) {
        this._elapsed = elapsed;
    }
    /**
     * Response result as JSON.
     */
    get result() {
        return this._result;
    }
    /**
     * Response result as JSON.
     */
    set result(result) {
        this._result = result;
    }
    /**
     * Request headers.
     */
    get headers() {
        return this._headers;
    }
    /**
     * Request headers.
     */
    set headers(headers) {
        this._headers = headers;
    }
    /**
     * Response HTTP status code.
     */
    get statusCode() {
        return this._statusCode;
    }
    /**
     * Response HTTP status code.
     */
    set statusCode(statusCode) {
        this._statusCode = statusCode;
    }
    /**
     * Indicates if request should be retried (forced).
     */
    get forceRetry() {
        return this._forceRetry;
    }
    /**
     * Indicates if request should be retried (forced).
     */
    set forceRetry(forceRetry) {
        this._forceRetry = forceRetry;
    }
    /**
     * Method used to check if request has errors.
     */
    requestHasErrors() {
        switch (this._statusCode) {
            case 0:
            case StatusCodes.Ok:
            case StatusCodes.Created:
            case StatusCodes.NonAuthoritativeInformation:
            case StatusCodes.NoContent:
            case StatusCodes.NotModified:
            case StatusCodes.NotFound: {
                return false;
            }
            default: {
                return true;
            }
        }
    }
}
