// ravendb/dist/commonjs/ServerWide/Operations/Analyzers/DeleteServerWideAnalyzerOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteServerWideAnalyzerOperation = void 0;
const index_js_1 = require("../../../Exceptions/index.js");
const RavenCommand_js_1 = require("../../../Http/RavenCommand.js");
const RaftIdGenerator_js_1 = require("../../../Utility/RaftIdGenerator.js");
class DeleteServerWideAnalyzerOperation {
    _analyzerName;
    constructor(analyzerName) {
        if (!analyzerName) {
            (0, index_js_1.throwError)("InvalidArgumentException", "AnalyzerName cannot be null");
        }
        this._analyzerName = analyzerName;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new DeleteServerWideAnalyzerCommand(this._analyzerName);
    }
}
exports.DeleteServerWideAnalyzerOperation = DeleteServerWideAnalyzerOperation;
class DeleteServerWideAnalyzerCommand extends RavenCommand_js_1.RavenCommand {
    _analyzerName;
    constructor(analyzerName) {
        super();
        if (!analyzerName) {
            (0, index_js_1.throwError)("InvalidArgumentException", "AnalyzerName cannot be null");
        }
        this._analyzerName = analyzerName;
    }
    createRequest(node) {
        const uri = node.url + "/admin/analyzers?name=" + this._urlEncode(this._analyzerName);
        return {
            uri,
            method: "DELETE"
        };
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator_js_1.RaftIdGenerator.newId();
    }
}
