// ravendb/dist/esm/ServerWide/Operations/Configuration/PutDatabaseSettingsOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class PutDatabaseSettingsOperation {
    _databaseName;
    _configurationSettings;
    constructor(databaseName, configurationSettings) {
        if (!databaseName) {
            throwError("InvalidArgumentException", "DatabaseName cannot be null");
        }
        this._databaseName = databaseName;
        if (!configurationSettings) {
            throwError("InvalidArgumentException", "ConfigurationSettings cannot be null");
        }
        this._configurationSettings = configurationSettings;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new PutDatabaseConfigurationSettingsCommand(this._configurationSettings, this._databaseName);
    }
}
class PutDatabaseConfigurationSettingsCommand extends RavenCommand {
    _configurationSettings;
    _databaseName;
    constructor(configurationSettings, databaseName) {
        super();
        if (!databaseName) {
            throwError("InvalidArgumentException", "DatabaseName cannot be null");
        }
        this._databaseName = databaseName;
        if (!configurationSettings) {
            throwError("InvalidArgumentException", "ConfigurationSettings cannot be null");
        }
        this._configurationSettings = configurationSettings;
    }
    get isReadRequest() {
        return false;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + this._databaseName + "/admin/configuration/settings";
        const body = this._serializer.serialize(this._configurationSettings);
        return {
            uri,
            method: "PUT",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
}
