// ravendb/dist/esm/ServerWide/Operations/Configuration/PutServerWideClientConfigurationOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class PutServerWideClientConfigurationOperation {
    _configuration;
    constructor(configuration) {
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get resultType() {
        return "CommandResult";
    }
    getCommand(conventions) {
        return new PutServerWideClientConfigurationCommand(conventions, this._configuration);
    }
}
class PutServerWideClientConfigurationCommand extends RavenCommand {
    _configuration;
    constructor(conventions, configuration) {
        super();
        if (!conventions) {
            throwError("InvalidArgumentException", "Conventions cannot be null");
        }
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null");
        }
        this._configuration = configuration;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/admin/configuration/client";
        const body = this._serializer.serialize(this._configuration);
        return {
            uri,
            method: "PUT",
            headers: this._headers().typeAppJson().build(),
            body
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
