// ravendb/dist/esm/Documents/Operations/CompareExchange/GetCompareExchangeValueOperation.js
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { throwError } from "../../../Exceptions/index.js";
import { CompareExchangeValueResultParser } from "./CompareExchangeValueResultParser.js";
import { GetCompareExchangeValuesCommand } from "./GetCompareExchangeValuesOperation.js";
export class GetCompareExchangeValueOperation {
    _key;
    _materializeMetadata;
    _clazz;
    constructor(key, clazz, materializeMetadata = true) {
        this._key = key;
        this._clazz = clazz;
        this._materializeMetadata = materializeMetadata;
    }
    getCommand(store, conventions, cache) {
        return new GetCompareExchangeValueCommand(this._key, this._materializeMetadata, conventions, this._clazz);
    }
    get resultType() {
        return "CommandResult";
    }
}
export class GetCompareExchangeValueCommand extends RavenCommand {
    _key;
    _clazz;
    _materializeMetadata;
    _conventions;
    constructor(key, materializeMetadata, conventions, clazz) {
        super();
        if (!key) {
            throwError("InvalidArgumentException", "The key argument must have value");
        }
        this._key = key;
        this._clazz = clazz;
        this._materializeMetadata = materializeMetadata;
        this._conventions = conventions;
    }
    get isReadRequest() {
        return true;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/cmpxchg?key=" + encodeURIComponent(this._key);
        return { uri };
    }
    async setResponseAsync(bodyStream, fromCache) {
        if (!bodyStream) {
            return null;
        }
        let body = null;
        const results = await this._pipeline()
            .collectBody(x => body = x)
            .parseJsonSync()
            .process(bodyStream);
        const localObject = GetCompareExchangeValuesCommand.mapToLocalObject(results);
        this.result = CompareExchangeValueResultParser.getValue(localObject, this._materializeMetadata, this._conventions, this._clazz);
        return body;
    }
}
