// ravendb/dist/esm/Documents/Operations/CompareExchange/PutCompareExchangeValueOperation.js
import { CompareExchangeResult } from "./CompareExchangeResult.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { DocumentConventions } from "../../Conventions/DocumentConventions.js";
import { throwError } from "../../../Exceptions/index.js";
import { JsonSerializer } from "../../../Mapping/Json/Serializer.js";
import { TypeUtil } from "../../../Utility/TypeUtil.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
import { COMPARE_EXCHANGE, CONSTANTS } from "../../../Constants.js";
import { CompareExchangeSessionValue } from "./CompareExchangeSessionValue.js";
export class PutCompareExchangeValueOperation {
    _key;
    _value;
    _index;
    _metadata;
    constructor(key, value, index, metadata) {
        this._key = key;
        this._value = value;
        this._index = index;
        this._metadata = metadata;
    }
    getCommand(store, conventions, cache) {
        return new PutCompareExchangeValueCommand(this._key, this._value, this._index, this._metadata, conventions);
    }
    get resultType() {
        return "CommandResult";
    }
}
export class PutCompareExchangeValueCommand extends RavenCommand {
    _key;
    _value;
    _index;
    _conventions;
    _metadata;
    constructor(key, value, index, metadata, conventions) {
        super();
        if (!key) {
            throwError("InvalidArgumentException", "The key argument must have value");
        }
        if (index < 0) {
            throwError("InvalidArgumentException", "Index must be a non-negative number");
        }
        this._key = key;
        this._value = value;
        this._index = index;
        this._metadata = metadata;
        this._conventions = conventions || DocumentConventions.defaultConventions;
    }
    get isReadRequest() {
        return false;
    }
    createRequest(node) {
        const uri = node.url + "/databases/" + node.database + "/cmpxchg?key=" + encodeURIComponent(this._key) + "&index=" + this._index;
        const tuple = {};
        tuple[COMPARE_EXCHANGE.OBJECT_FIELD_NAME] = TypeUtil.isPrimitive(this._value)
            ? this._value
            : this._conventions.transformObjectKeysToRemoteFieldNameConvention(this._value);
        if (this._metadata) {
            const metadata = CompareExchangeSessionValue.prepareMetadataForPut(this._key, this._metadata, this._conventions);
            tuple[CONSTANTS.Documents.Metadata.KEY] = metadata;
        }
        return {
            method: "PUT",
            uri,
            body: JsonSerializer.getDefault().serialize(tuple),
            headers: this._headers().typeAppJson().build()
        };
    }
    async setResponseAsync(bodyStream, fromCache) {
        let body = null;
        const resObj = await this._pipeline()
            .collectBody(_ => body = _)
            .parseJsonSync()
            .process(bodyStream);
        const type = !TypeUtil.isPrimitive(this._value)
            ? this._conventions.getTypeDescriptorByEntity(this._value)
            : null;
        const clazz = TypeUtil.isClass(type) ? type : null;
        this.result = CompareExchangeResult.parseFromObject(resObj, this._conventions, clazz);
        return body;
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
