// ravendb/dist/esm/Documents/Operations/Configuration/PutClientConfigurationOperation.js
import { throwError } from "../../../Exceptions/index.js";
import { RavenCommand } from "../../../Http/RavenCommand.js";
import { HeadersBuilder } from "../../../Utility/HttpUtil.js";
import { RaftIdGenerator } from "../../../Utility/RaftIdGenerator.js";
export class PutClientConfigurationOperation {
    _configuration;
    get resultType() {
        return "CommandResult";
    }
    constructor(configuration) {
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null or undefined.");
        }
        this._configuration = configuration;
    }
    getCommand(conventions) {
        return new PutClientConfigurationCommand(conventions, this._configuration);
    }
}
export class PutClientConfigurationCommand extends RavenCommand {
    _configuration;
    get isReadRequest() {
        return false;
    }
    get responseType() {
        return "Empty";
    }
    constructor(conventions, configuration) {
        super();
        if (!conventions) {
            throwError("InvalidArgumentException", "Document conventions cannot be null or undefined.");
        }
        if (!configuration) {
            throwError("InvalidArgumentException", "Configuration cannot be null or undefined.");
        }
        this._configuration = this._serializer.serialize(configuration);
    }
    createRequest(node) {
        const uri = `${node.url}/databases/${node.database}/admin/configuration/client`;
        return {
            method: "PUT",
            uri,
            body: this._configuration,
            headers: HeadersBuilder.create()
                .typeAppJson()
                .build()
        };
    }
    getRaftUniqueRequestId() {
        return RaftIdGenerator.newId();
    }
}
